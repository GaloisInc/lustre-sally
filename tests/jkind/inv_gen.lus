-- This file contains a simple property which requires the invariant
-- described by 'lemma' (either explicitly or through invariant
-- generation)

node main() returns (x: bool);
var
   count: int; 
   ok: bool;
   lemma: bool;
let
   count = 1 -> 
      (if (pre(count) < 5)
         then pre(count) + 1
         else 5);

   x = false -> ((not (count >= 5)) and (pre x));

   ok = true -> ((not (pre x)) or x);
   --%PROPERTY ok;

   lemma = not x;
   -- %PROPERTY lemma;
tel;
