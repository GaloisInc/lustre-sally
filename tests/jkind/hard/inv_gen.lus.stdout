Core Lustre
===========

node ()
  returns (x)
  assumes ()
  shows (ok)
let
  valof_pre_count_5 = pre count
  valof_pre_count_6 = pre count
  count = 1 -> ITE(Lt(valof_pre_count_5, 5), Add(valof_pre_count_6, 1), 5)
  valof_pre_x_7 = pre x
  x = false -> And(Not(Geq(count, 5)), valof_pre_x_7)
  valof_pre_x_8 = pre x
  ok = true -> Or(Not(valof_pre_x_8), x)
  lemma = Not(x)
tel
Sally Model
===========

(define-state-type
   S
   (
     (count Int)
     (lemma Bool)
     (ok Bool)
     (x Bool)
     (|count:init| Bool)
     (|gal-initializing| Bool)
     (|ok:init| Bool)
     (|valof_pre_count:5| Int)
     (|valof_pre_count:6| Int)
     (|valof_pre_x:7| Bool)
     (|valof_pre_x:8| Bool)
     (|x:init| Bool))
   ())
(define-transition-system
   TS
   S
   (and
      (= |gal-initializing| true)
      (= |count:init| true)
      (= |x:init| true)
      (= |ok:init| true))
   (and
      (= next.|gal-initializing| false)
      (= next.|valof_pre_count:5| state.count)
      (= next.|valof_pre_count:6| state.count)
      (ite
         state.|count:init|
         (= next.count 1)
         (=
            next.count
            (ite
               (< next.|valof_pre_count:5| 5)
               (+ next.|valof_pre_count:6| 1)
               5)))
      (= next.|count:init| false)
      (= next.|valof_pre_x:7| state.x)
      (ite
         state.|x:init|
         (= next.x false)
         (=
            next.x
            (and
               (not
                  (<= 5 next.count))
               next.|valof_pre_x:7|)))
      (= next.|x:init| false)
      (= next.|valof_pre_x:8| state.x)
      (ite
         state.|ok:init|
         (= next.ok true)
         (=
            next.ok
            (or
               (not next.|valof_pre_x:8|)
               next.x)))
      (= next.|ok:init| false)
      (=
         next.lemma
         (not next.x))))

(query
   TS
   (or |gal-initializing| ok))

[Lustre] Validating properties:
[Lustre] Property ok... considering simultaneous states to depth 0
considering simultaneous states to depth 0
considering simultaneous states to depth 1
considering simultaneous states to depth 1
considering simultaneous states to depth 2
considering simultaneous states to depth 2
considering simultaneous states to depth 3
considering simultaneous states to depth 3
considering simultaneous states to depth 4
considering simultaneous states to depth 4
considering simultaneous states to depth 5
considering simultaneous states to depth 5
considering simultaneous states to depth 6
considering simultaneous states to depth 6
considering simultaneous states to depth 7
considering simultaneous states to depth 7
considering simultaneous states to depth 8
considering simultaneous states to depth 8
considering simultaneous states to depth 9
considering simultaneous states to depth 9
considering simultaneous states to depth 10
considering simultaneous states to depth 10
counter-example search depth 0
counter-example search depth 1
counter-example search depth 2
counter-example search depth 3
counter-example search depth 4
counter-example search depth 5
counter-example search depth 6
counter-example search depth 7
counter-example search depth 8
counter-example search depth 9
counter-example search depth 10
[[1;93mUnknown[0m] Valid up to depth 10
[Lustre] Model status: [[1;93mUnknown[0m] 
